internal/venice-client-common/src/main/java/org/apache/avro/generic/DeterministicMapOrderDatumWriter.java

```diff
@@ -18,16 +19,36 @@ public interface DeterministicMapOrderDatumWriter {
 
   default void writeMapWithDeterministicOrder(Schema schema, Object datum, Encoder out) throws IOException {
     Schema valueSchemaType = schema.getValueType();
-    Map map = (Map) datum;
+    Map<? extends CharSequence, Object> map = (Map<? extends CharSequence, Object>) datum;
     final int expectedMapSize = map.size();
     int actualSize = 0;
     out.writeMapStart();
     out.setItemCount(expectedMapSize);
 
-    List<Map.Entry> sortedEntryList =
-        (List<Map.Entry>) map.entrySet().stream().sorted(Map.Entry.comparingByKey()).collect(Collectors.toList());
+    List<Map.Entry<? extends CharSequence, Object>> sortedEntryList = map.entrySet().stream().sorted((e1, e2) -> {
```

Comment:
This seems like a hot path function. I don't think we should use Java streams in hot paths, as they generate a lot of garbage. I think we could rewrite this code like this:

```
    Map.Entry[] contentArray = map.entrySet().toArray(new Map.Entry[expectedMapSize]);
    Arrays.sort(contentArray, COMPARATOR); // COMPARATOR is a static property containing your algorithm below
    List<Map.Entry> sortedEntryList = Arrays.asList(contentArray);
```

Reply:
Yeah, that's a good suggestion. I replaced it with:
```
List<Map.Entry<? extends CharSequence, Object>> sortedEntryList = new ArrayList<>(map.entrySet());
sortedEntryList.sort(COMPARATOR);
```

I believe this incurs the same amount of data copy with 1 less LOC

Reply:
Hmm... I think that `ArrayList` constructor will result in two copies:

```
    public ArrayList(Collection<? extends E> c) {
        Object[] a = c.toArray();
        if ((size = a.length) != 0) {
            if (c.getClass() == ArrayList.class) {
                elementData = a;
            } else {
                elementData = Arrays.copyOf(a, size, Object[].class);
            }
        } else {
            // replace with empty array.
            elementData = EMPTY_ELEMENTDATA;
        }
    }
```

First, `toArray` will do a copy, and then `Arrays.copyOf` will do another one...

Whereas `Arrays.asList` does not do an extra copy...

Reply:
I see. Yeah, then I'll replace with your suggestion